astropy/io/fits/hdu/compressed/compressed.py

```diff
@@ -558,7 +582,10 @@ def _scale_data(self, data):
                 np.add(data, self._orig_bzero, out=data, casting="unsafe")
 
             if blanks is not None:
-                data = np.where(blanks, np.nan, data)
+                # use float32 version of nan to reduce data size for uint conversion
+                # result will still be float64 for larger uint sizes (e.g., uint32)
+                # for float types np.where retains the type of the data array
+                data = np.where(blanks, np.float32(np.nan), data)
```

Comment:
All this file is being heavily refactored in #15474, and things should be cleaner after since it should inherit more of ImageHDU's behavior... and probably get the float/NaN conversion for free.

Reply:
I don't doubt that (and certainly hope it is true!)  There might still be some work needed in the #15747  branch to handle unsigned ints consistently.  They are currently not handled correctly in `ImageHDU`.  So chances are you will inherit that incorrect behavior without some additional work on the `ImageHDU` class.

Reply:
Yes; I'd actually argue for a follow-up PR in this branch to make the corresponding fixes in `image.py`, so 6.1.x will have consistent behaviour of uncompressed files, and #15747 can inherit the fixed code right away.